lisa/core/node.py

```diff
@@ -35,17 +41,55 @@ def createNode(
 
     def setConnectionInfo(
         self,
-        address: str = "",
-        port: int = 22,
-        publicAddress: str = "",
-        publicPort: int = 22,
-        username: str = "root",
-        password: str = "",
-        privateKeyFile: str = "",
-    ):
-        self.connection = SshConnection(
-            address, port, publicAddress, publicPort, username, password, privateKeyFile
-        )
+        address: Optional[object],
+        port: Optional[object],
+        publicAddress: Optional[object],
+        publicPort: Optional[object],
+        username: Optional[object],
+        password: Optional[object],
+        privateKeyFile: Optional[object],
+    ) -> None:
+        if self.connection is not None:
+            raise Exception(
+                "node is set connection information already, cannot set again"
+            )
+        parameters: Dict[str, object] = dict()
+        if address is not None:
+            parameters["address"] = address
+        if port is not None:
+            parameters["port"] = port
+        if publicAddress is not None:
+            parameters["publicAddress"] = publicAddress
+        if publicPort is not None:
+            parameters["publicPort"] = publicPort
+        if username is not None:
+            parameters["username"] = username
+        if password is not None:
+            parameters["password"] = password
+        if privateKeyFile is not None:
+            parameters["privateKeyFile"] = privateKeyFile
+        self.connection = SshConnection(**parameters)
+
+    def execute(self, cmd: str) -> ExecutableResult:
+        result: ExecutableResult
+        cmd_id = random.randint(0, 10000)
+        start_timer = timer()
+        log.debug("remote(%s) cmd[%s] %s", self.isRemote, cmd_id, cmd)
+        if self.isRemote:
```

Comment:
yes, it can be. I haven't created subclasses for non-extendable features.

Reply:
I think this an instance where applying OO actually makes a lot of sense. By non-extendable do you mean we don't expect users to subclass `Node` themselves?

Reply:
I don't expect user to extend Node, so that Node is not split for Remote and Local. As it also handles Windows, Linux. We many need something split outside of Node. Like [Local, Remote][Process, File], [Windows, Linux]OS.

Reply:
Might be a good candidate for mix-ins, so all the implementation details specific to each kind remain separate, and we just compose a Node with the pieces we need.